### ISSUE TITLE
Custom User Fields Mapping Issue During Login

### ISSUE THREAD

**Issue by AANP-ITSupport** [flipboxfactory/saml-sp#5]
I created a custom User Field in Craft CMS, called memberType.  

After a little debugging of [/services/Login.php](https://github.com/flipboxfactory/saml-sp/blob/394509af9129d8bbda362fb778202a235c38d6d6/src/services/Login.php), and some major confusion, I now understand that this PHP function – [property_exists](https://stackoverflow.com/questions/16729830/check-if-a-property-exists-on-magically-set-properties), is returning false for these custom User Fields since they are dynamically added to the model.  Even though I can get and set these fields from $user->memberType, for example.  Therefore, the simple mapping logic doesn’t apply.
 
Because property_exists is returning false, it falls into the “[look for a callable](https://github.com/flipboxfactory/saml-sp/blob/0f12071e98d02d4a68e50f615f7e5f9ccfeeca41/src/services/Login.php#L367)” section .
 
So, I thought I would attempt to create a callable, but when I add this to the responseAttributeMap it fails with “Constant expression contains invalid operations”.    It sounds to me like this property is not allowing functions to be added to the array, but the examples given seem to be doing just that?
 
My knowledge of PHP is about as basic as you can get.  So I’m sure I’m missing something simple here.

![custom_user_fields_loginphp](https://user-images.githubusercontent.com/18406934/40387584-c9408b50-5dd2-11e8-8aa8-bf44d1bdfe30.jpg)

**Comment by dsmrt** [comment]
Yep ... looks like that callable section is getting skipped over. I'll get a patch going for this.

**Comment by dsmrt** [comment]
Fixed in 1.0.0-beta.9.
I still need to add some documentation on this but the callback will pass the user then the attribute. For example, your `configs/saml-sp.php` can look like the following:
```php
<?php

return [
    'responseAttributeMap' => [
        \LightSaml\ClaimTypes::EMAIL_ADDRESS => 'email',
        \LightSaml\ClaimTypes::GIVEN_NAME    => 'firstName',
        \LightSaml\ClaimTypes::SURNAME       => function(\craft\elements\User $user,
                                                         \LightSaml\Model\Assertion\Attribute $attribute) {
            $user->lastName = $attribute->getFirstAttributeValue();
        }
    ],
];
```
